crates/ty_python_semantic/src/module_resolver/list.rs

```diff
@@ -0,0 +1,1688 @@
+    use ruff_db::system::{
+        DbWithTestSystem, DbWithWritableSystem, OsSystem, SystemPath, SystemPathBuf,
+    };
+    use ruff_db::testing::assert_function_query_was_not_run;
+    use ruff_python_ast::PythonVersion;
+
+    use crate::db::{Db, tests::TestDb};
+    use crate::module_resolver::module::Module;
+    use crate::module_resolver::resolver::{
+        ModuleResolveMode, ModuleResolveModeIngredient, dynamic_resolution_paths,
+    };
+    use crate::module_resolver::testing::{FileSpec, MockedTypeshed, TestCase, TestCaseBuilder};
+    use crate::program::{Program, ProgramSettings, SearchPathSettings};
+    use crate::{PythonPlatform, PythonVersionSource, PythonVersionWithSource};
+
+    use super::list_modules;
+
+    struct ModuleDebugSnapshot<'db> {
+        db: &'db dyn Db,
+        module: Module<'db>,
+    }
+
+    impl std::fmt::Debug for ModuleDebugSnapshot<'_> {
+        fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+            match self.module {
+                Module::Namespace(pkg) => {
+                    write!(f, "Module::Namespace({name:?})", name = pkg.name(self.db))
+                }
+                Module::File(module) => {
+                    // For snapshots, just normalize all paths to using
+                    // Unix slashes for simplicity.
+                    let path_components = match module.file(self.db).path(self.db) {
+                        FilePath::System(path) => path.as_path().components(),
+                        FilePath::Vendored(path) => path.as_path().components(),
+                        FilePath::SystemVirtual(path) => Utf8Path::new(path.as_str()).components(),
+                    };
+                    let nice_path = path_components
+                        .map(|component| component.as_str())
+                        .collect::<Vec<&str>>()
+                        .join("/");
+                    write!(
+                        f,
+                        "Module::File({name:?}, {search_path:?}, {path:?}, {kind:?}, {known:?})",
+                        name = module.name(self.db).as_str(),
+                        search_path = module.search_path(self.db).debug_kind(),
+                        path = nice_path,
+                        kind = module.kind(self.db),
+                        known = module.known(self.db),
+                    )
+                }
+            }
+        }
+    }
+
+    fn sorted_list(db: &dyn Db) -> Vec<Module<'_>> {
+        let mut modules = list_modules(db);
+        modules.sort_by(|m1, m2| {
+            let key1 = (
+                m1.name(db),
+                m1.search_path(db).map(|sp| sp.as_path()),
+                m1.kind(db),
+                m1.known(db),
+            );
+            let key2 = (
+                m2.name(db),
+                m2.search_path(db).map(|sp| sp.as_path()),
+                m2.kind(db),
+                m2.known(db),
+            );
+            key1.cmp(&key2)
+        });
```

Comment:
Shouldn't it be enough to sort by module name? I don't think we should ever return multiple modules with the same name?